File: 3382-find-the-number-of-subarrays-where-boundary-elements-are-maximum/find-the-number-of-subarrays-where-boundary-elements-are-maximum.java
class Solution {
    public long numberOfSubarrays(int[] nums) {
        int n = nums.length;
        long ans = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        Stack<Integer> st = new Stack<>();
        for(int i = 0; i < n; i++) {
            while(!st.isEmpty() && nums[st.peek()] < nums[i]) {
                map.put(nums[st.peek()], map.get(nums[st.peek()])-1);
                st.pop();
            }
            st.push(i);
            map.put(nums[i], map.getOrDefault(nums[i], 0)+1);
            ans += map.get(nums[i]);
        }
        return ans;
    }
}

File: 3382-find-the-number-of-subarrays-where-boundary-elements-are-maximum/find-the-number-of-subarrays-where-boundary-elements-are-maximum.cpp
class Solution {
public:
    long long numberOfSubarrays(vector<int>& nums) {
        unordered_map<int, int> map;
        stack<int> st;

        int n = nums.size();
        long long ans = 0;

        for(int i = 0; i < n; i++) {
            while(!st.empty() && nums[st.top()] < nums[i]) {
                if(map.count(nums[st.top()])) {
                    map.erase(nums[st.top()]);
                }
                st.pop();
            }
            
            st.push(i);
            int val = map[nums[i]];
            map[nums[i]] = val + 1;
            ans += val+1;
        }

        return ans;
    }
};